// @guanjia-openclaw/report/src/server/install-reporter.ts
/**
 * InstallReporter - 安装上报模块
 *
 * 在 Electron 主进程中检测安装状态并上报，兼容 Windows 和 macOS。
 * 通过标记文件（.installed）判断是首次安装还是升级。
 *
 * 上报时机：
 * - 检测到安装（首次安装或升级）后启动 → expo (action_type=Install)
 * - 卸载时 → expo (action_type=Uninstall)
 * - 正常启动（版本相同）→ 不上报
 */

import { existsSync, readFileSync, writeFileSync, mkdirSync } from 'node:fs'
import { join } from 'node:path'
import { ServerAction } from './action.js'
import {
  DEFAULT_APPKEY,
  REPORT_CONST,
  getReportPlatformId,
} from '../constants.js'
import type { InstallChannel } from '../constants.js'

/** 标记文件名 */
const INSTALL_MARKER_FILE = '.installed'

/** 标记文件结构 */
interface InstallMarker {
  /** 上次安装的版本号 */
  version: string
  /** 安装时间（ISO 格式） */
  installedAt: string
  /** 安装平台 */
  platform: string
}

export interface InstallReporterOptions {
  /** 状态目录路径（如 ~/.qclaw），标记文件将写入该目录 */
  stateDir: string
  /** 当前应用版本号 */
  appVersion: string
  /** 上报地址 */
  baseUrl: string;
  /** 设备唯一标识（machineId） */
  guid: string
  /** 安装渠道号（如 5001=官网 / 5002=内测群） */
  channel?: InstallChannel
  /** 自定义 AppKey，默认 'PC_Qclaw' */
  appkey?: string
  /** 日志输出函数（可选，默认 console） */
  logger?: {
    info: (msg: string, ...args: unknown[]) => void
    warn: (msg: string, ...args: unknown[]) => void
  }
}

export class InstallReporter {
  private stateDir: string
  private appVersion: string
  private channel: InstallChannel | undefined
  private action: ServerAction
  private baseUrl: string
  private logger: NonNullable<InstallReporterOptions['logger']>

  constructor(options: InstallReporterOptions) {
    this.stateDir = options.stateDir
    this.appVersion = options.appVersion
    this.channel = options.channel
    this.baseUrl = options.baseUrl
    this.logger = options.logger ?? {
      info: console.log,
      warn: console.warn,
    }

    this.action = new ServerAction({
      appkey: options.appkey ?? DEFAULT_APPKEY,
      guid: options.guid,
      baseUrl: options.baseUrl,
      onReportSuccess: () => {
        this.logger.info('[InstallReporter] 上报成功')
      },
      onReportFail: (retCode) => {
        this.logger.warn(`[InstallReporter] 上报失败, retCode: ${retCode}`)
      },
    })
  }

  /**
   * 获取标记文件完整路径
   */
  private getMarkerPath(): string {
    return join(this.stateDir, INSTALL_MARKER_FILE)
  }

  /**
   * 读取标记文件
   */
  private readMarker(): InstallMarker | null {
    try {
      const markerPath = this.getMarkerPath()
      if (!existsSync(markerPath)) return null
      const content = readFileSync(markerPath, 'utf-8')
      return JSON.parse(content) as InstallMarker
    } catch {
      return null
    }
  }

  /**
   * 写入标记文件
   */
  private writeMarker(): void {
    try {
      if (!existsSync(this.stateDir)) {
        mkdirSync(this.stateDir, { recursive: true })
      }
      const marker: InstallMarker = {
        version: this.appVersion,
        installedAt: new Date().toISOString(),
        platform: process.platform,
      }
      writeFileSync(this.getMarkerPath(), JSON.stringify(marker, null, 2), 'utf-8')
    } catch (err) {
      this.logger.warn('[InstallReporter] 写入标记文件失败:', err)
    }
  }

  /**
   * 构建通用上报参数
   */
  private buildCommonParams(): Record<string, string> {
    return {
      channel: String(this.channel ?? ''),
      app_version: this.appVersion,
      tdbank_imp_date: Date.now().toString(),
      event_params: getReportPlatformId(),
    }
  }

  /**
   * 检测安装状态并上报
   *
   * - 标记文件不存在或版本号不同 → 检测到安装，同时上报 expo 和 click_new 事件（action_type=Install）
   * - 标记文件版本号相同 → 正常启动，不上报
   *
   * @returns 上报是否成功（正常启动不上报时返回 true）
   */
  async checkAndReport(): Promise<boolean> {
    try {
      const marker = this.readMarker()
      const commonParams = this.buildCommonParams()

      if (!marker || marker.version !== this.appVersion) {
        // 检测到安装（首次安装或升级），统一上报
        const previousVersion = marker?.version
        this.logger.info(
          `[InstallReporter] 检测到安装 v${this.appVersion}${
            previousVersion ? `（从 v${previousVersion} 升级）` : '（首次安装）'
          }`,
        )
        const installParams = {
          page_id: 'Install_Page',
          action_type: 'Install',
          ...commonParams,
        }
        // 同时上报 expo 和 click_new 两个事件
        const [expoOk, clickOk] = await Promise.all([
          this.action.reportAsync(REPORT_CONST.EXPO, installParams),
          this.action.reportAsync(REPORT_CONST.CLICK_NEW, installParams),
        ])
        this.writeMarker()
        return expoOk && clickOk
      }

      // 版本号相同：正常启动，不上报
      return true
    } catch (err) {
      // 安装上报不应影响应用正常启动
      this.logger.warn('[InstallReporter] 安装上报异常:', err)
      return false
    }
  }

  /**
   * 上报卸载曝光事件（expo）
   *
   * 卸载页面展示时调用一次，用于记录卸载页面曝光。
   * Windows 端由 NSIS 卸载脚本调用，macOS 端由应用退出事件触发。
   *
   * @returns 上报是否成功
   */
  async reportUninstallExpo(): Promise<boolean> {
    try {
      this.logger.info(`[InstallReporter] 上报卸载曝光(expo) v${this.appVersion}`)
      const commonParams = this.buildCommonParams()
      return await this.action.reportAsync(REPORT_CONST.EXPO, {
        page_id: 'Uninstall_Page',
        action_type: 'Uninstall',
        ...commonParams,
      })
    } catch (err) {
      this.logger.warn('[InstallReporter] 卸载曝光上报异常:', err)
      return false
    }
  }

  /**
   * 上报卸载点击事件（click_new）
   *
   * 用户在卸载过程中每次点击操作时调用，用于追踪用户卸载行为。
   * Windows 端由 NSIS 卸载脚本调用，macOS 端由应用退出事件触发。
   *
   * @returns 上报是否成功
   */
  async reportUninstallClick(): Promise<boolean> {
    try {
      this.logger.info(`[InstallReporter] 上报卸载点击(click_new) v${this.appVersion}`)
      const commonParams = this.buildCommonParams()
      return await this.action.reportAsync(REPORT_CONST.CLICK_NEW, {
        page_id: 'Uninstall_Page',
        action_type: 'Uninstall',
        ...commonParams,
      })
    } catch (err) {
      this.logger.warn('[InstallReporter] 卸载点击上报异常:', err)
      return false
    }
  }

  /**
   * 上报卸载事件（兼容方法，同时上报 expo + click_new）
   *
   * @deprecated 请使用 reportUninstallExpo() 和 reportUninstallClick() 分别调用
   * @returns 上报是否成功
   */
  async reportUninstall(): Promise<boolean> {
    const [expoOk, clickOk] = await Promise.all([
      this.reportUninstallExpo(),
      this.reportUninstallClick(),
    ])
    return expoOk && clickOk
  }
}
