// @guanjia-openclaw/report/src/constants.ts
/**
 * 上报模块 - 公共常量
 *
 * 客户端（浏览器）和服务端（Node.js）共用的常量定义。
 */

// ==================== 上报地址 ====================

/** 正式环境上报地址 */
export const REPORT_URL_PRODUCTION = 'https://pcmgrmonitor.3g.qq.com/datareport'

/** 测试环境上报地址 */
export const REPORT_URL_TEST = 'https://pcmgrmonitor.3g.qq.com/test/datareport'

/** 本地开发代理地址（Vite proxy → 测试环境） */
export const REPORT_URL_LOCAL_PROXY = '/beacon/test/datareport'

/** 默认 AppKey */
export const DEFAULT_APPKEY = 'PC_Qclaw'

// ==================== 上报事件代码 ====================

/** 上报事件常量 */
export const REPORT_CONST = {
  /** 点击事件 */
  CLICK_NEW: 'click_new',
  /** 曝光事件 */
  EXPO: 'expo',
} as const

export type ReportConstType = (typeof REPORT_CONST)[keyof typeof REPORT_CONST]

// ==================== 安装事件代码 ====================

/** 首次安装成功（应用首次启动时上报） */
export const INSTALL_EVENT_FIRST_INSTALL = 'install_success'

/** 卸载成功 */
export const INSTALL_EVENT_UNINSTALL = 'uninstall_success'

/** NSIS 安装开始（Windows 安装过程中上报） */
export const INSTALL_EVENT_NSIS_START = 'nsis_install_start'

/** NSIS 安装完成（Windows 安装过程中上报） */
export const INSTALL_EVENT_NSIS_FINISH = 'nsis_install_finish'

/** NSIS 安装取消（Windows 安装过程中上报） */
export const INSTALL_EVENT_NSIS_CANCEL = 'nsis_install_cancel'

// ==================== 平台标识 ====================

/**
 * 获取上报用的平台标识（event_params 字段）
 *
 * - Windows → Qclaw_Win
 * - macOS ARM → Qclaw_MAC_ARM
 * - macOS Intel → Qclaw_MAC_INTEL
 */
export function getReportPlatformId(): string {
  if (process.platform === 'win32') {
    return 'Qclaw_Win'
  }
  if (process.platform === 'darwin') {
    return process.arch === 'arm64' ? 'Qclaw_MAC_ARM' : 'Qclaw_MAC_INTEL'
  }
  return `Qclaw_${process.platform}_${process.arch}`
}

// ==================== 环境类型 ====================

export const ENV_TYPES = {
  LOCAL: 'test',
  PRODUCTION: 'production',
} as const

export type EnvType = (typeof ENV_TYPES)[keyof typeof ENV_TYPES]

// ==================== 安装渠道 ====================

/** 安装渠道类型 */
export const INSTALL_CHANNELS = {
  /** 官网 */
  OFFICIAL: 5001,
  /** 内测群 */
  BETA_GROUP: 5002,
} as const

export type InstallChannel = (typeof INSTALL_CHANNELS)[keyof typeof INSTALL_CHANNELS]
