// @guanjia-openclaw/report/dist/client/action.js
/**
 * ClientAction - 客户端埋点上报
 *
 * 浏览器环境专用，依赖 fetch / navigator.sendBeacon。
 * 实现 IReportAction 接口，与 ServerAction 保持一致的调用方式。
 */
import { DEFAULT_APPKEY } from '../constants.js';
export class ClientAction {
    appkey;
    guid;
    onReportSuccess;
    onReportFail;
    baseUrl;
    constructor(options) {
        if (!options.baseUrl) {
            throw new Error('[ClientAction] baseUrl 为必填项，请从上层传入上报地址');
        }
        this.baseUrl = options.baseUrl;
        this.appkey = options.appkey ?? DEFAULT_APPKEY;
        this.guid = options.guid ?? '';
        this.onReportSuccess = options.onReportSuccess ?? (() => { });
        this.onReportFail = options.onReportFail ?? (() => { });
    }
    /**
     * 构建上报 payload
     */
    buildPayload(eventCode, eventParams, config) {
        const normalized = {};
        for (const key in eventParams) {
            const val = eventParams[key];
            normalized[key] = val != null ? String(val) : '';
        }
        normalized['tdbank_imp_date'] = String(Date.now());
        const params = {
            app_key: this.appkey,
            version: '1.0.0',
            guid: this.guid,
            event_code: eventCode,
            params: normalized,
        };
        if (config) {
            for (const key in config) {
                params[key] = config[key];
            }
        }
        return params;
    }
    /**
     * 立即上报（fire-and-forget）
     */
    reportDirect(eventCode, eventParams, config) {
        const params = this.buildPayload(eventCode, eventParams, config);
        fetch(this.baseUrl, {
            headers: { 'Content-Type': 'application/json' },
            method: 'POST',
            body: JSON.stringify(params),
            mode: 'cors',
        })
            .then((response) => response.json())
            .then((data) => {
            const result = data;
            if (result['retCode'] === 0) {
                this.onReportSuccess();
            }
            else {
                this.onReportFail(result['retCode']);
            }
        })
            .catch(() => this.onReportFail());
    }
    /**
     * 异步上报（返回 Promise）
     */
    async reportAsync(eventCode, eventParams, config) {
        const params = this.buildPayload(eventCode, eventParams, config);
        try {
            const response = await fetch(this.baseUrl, {
                headers: { 'Content-Type': 'application/json' },
                method: 'POST',
                body: JSON.stringify(params),
                mode: 'cors',
            });
            const data = (await response.json());
            if (data['retCode'] === 0) {
                this.onReportSuccess();
                return true;
            }
            this.onReportFail(data['retCode']);
            return false;
        }
        catch {
            this.onReportFail();
            return false;
        }
    }
    /**
     * 保守上报（页面关闭/跳转场景，使用 sendBeacon 避免丢失）
     *
     * 仅客户端支持，服务端无此方法。
     */
    reportConservation(eventCode, eventParams, config) {
        const params = this.buildPayload(eventCode, eventParams, config);
        navigator.sendBeacon(this.baseUrl, JSON.stringify(params));
    }
}
