// @guanjia-openclaw/report/src/types.ts
/**
 * 上报模块 - 公共类型定义
 *
 * 定义客户端和服务端共用的接口。
 */

import type { EnvType } from './constants.js'

/** 上报成功回调 */
export type ReportCallback = () => void

/** 上报失败回调 */
export type ReportFailCallback = (retCode?: number) => void

/**
 * 上报实例的统一接口
 *
 * 客户端 Action 和服务端 ServerAction 都实现此接口，
 * 保证 reportFunc 可以无差别调用。
 */
export interface IReportAction {
  /**
   * 立即上报（fire-and-forget）
   */
  reportDirect(
    eventCode: string,
    eventParams: Record<string, string | number>,
    config?: Record<string, unknown>,
  ): void

  /**
   * 异步上报（返回 Promise）
   */
  reportAsync(
    eventCode: string,
    eventParams: Record<string, string | number>,
    config?: Record<string, unknown>,
  ): Promise<boolean>
}

/**
 * 统一上报初始化选项
 */
export interface ReportInitOptions {
  /** 应用标识，默认 'PC_Qclaw' */
  appkey?: string
  /** 用户/设备唯一标识 */
  guid?: string
  /** 当前环境 */
  env?: EnvType
  /** 自定义上报地址（优先级高于 env 推导的地址） */
  baseUrl?: string
  /** 上报成功回调 */
  onReportSuccess?: ReportCallback
  /** 上报失败回调 */
  onReportFail?: ReportFailCallback
}
