// @guanjia-openclaw/report/src/client/action.ts
/**
 * ClientAction - 客户端埋点上报
 *
 * 浏览器环境专用，依赖 fetch / navigator.sendBeacon。
 * 实现 IReportAction 接口，与 ServerAction 保持一致的调用方式。
 */

import { DEFAULT_APPKEY } from '../constants.js'
import type { IReportAction, ReportCallback, ReportFailCallback, ReportInitOptions } from '../types.js'

// 浏览器环境类型声明（避免全局引入 DOM lib 影响服务端代码）
declare const navigator: { sendBeacon(url: string, data?: string): boolean }
declare function fetch(input: string, init?: Record<string, unknown>): Promise<{ json(): Promise<unknown> }>

export class ClientAction implements IReportAction {
  private appkey: string
  private guid: string
  private onReportSuccess: ReportCallback
  private onReportFail: ReportFailCallback
  private baseUrl: string

  constructor(options: ReportInitOptions) {
    if (!options.baseUrl) {
      throw new Error('[ClientAction] baseUrl 为必填项，请从上层传入上报地址')
    }
    this.baseUrl = options.baseUrl
    this.appkey = options.appkey ?? DEFAULT_APPKEY
    this.guid = options.guid ?? ''
    this.onReportSuccess = options.onReportSuccess ?? (() => {})
    this.onReportFail = options.onReportFail ?? (() => {})
  }

  /**
   * 构建上报 payload
   */
  private buildPayload(
    eventCode: string,
    eventParams: Record<string, string | number>,
    config?: Record<string, unknown>,
  ): Record<string, unknown> {
    const normalized: Record<string, string> = {}
    for (const key in eventParams) {
      const val = eventParams[key]
      normalized[key] = val != null ? String(val) : ''
    }
    normalized['tdbank_imp_date'] = String(Date.now())

    const params: Record<string, unknown> = {
      app_key: this.appkey,
      version: '1.0.0',
      guid: this.guid,
      event_code: eventCode,
      params: normalized,
    }

    if (config) {
      for (const key in config) {
        params[key] = config[key]
      }
    }

    return params
  }

  /**
   * 立即上报（fire-and-forget）
   */
  reportDirect(
    eventCode: string,
    eventParams: Record<string, string | number>,
    config?: Record<string, unknown>,
  ): void {
    const params = this.buildPayload(eventCode, eventParams, config)

    fetch(this.baseUrl, {
      headers: { 'Content-Type': 'application/json' },
      method: 'POST',
      body: JSON.stringify(params),
      mode: 'cors',
    })
      .then((response) => response.json())
      .then((data) => {
        const result = data as Record<string, unknown>
        if (result['retCode'] === 0) {
          this.onReportSuccess()
        } else {
          this.onReportFail(result['retCode'] as number)
        }
      })
      .catch(() => this.onReportFail())
  }

  /**
   * 异步上报（返回 Promise）
   */
  async reportAsync(
    eventCode: string,
    eventParams: Record<string, string | number>,
    config?: Record<string, unknown>,
  ): Promise<boolean> {
    const params = this.buildPayload(eventCode, eventParams, config)

    try {
      const response = await fetch(this.baseUrl, {
        headers: { 'Content-Type': 'application/json' },
        method: 'POST',
        body: JSON.stringify(params),
        mode: 'cors',
      })
      const data = (await response.json()) as Record<string, unknown>
      if (data['retCode'] === 0) {
        this.onReportSuccess()
        return true
      }
      this.onReportFail(data['retCode'] as number)
      return false
    } catch {
      this.onReportFail()
      return false
    }
  }

  /**
   * 保守上报（页面关闭/跳转场景，使用 sendBeacon 避免丢失）
   *
   * 仅客户端支持，服务端无此方法。
   */
  reportConservation(
    eventCode: string,
    eventParams: Record<string, string | number>,
    config?: Record<string, unknown>,
  ): void {
    const params = this.buildPayload(eventCode, eventParams, config)
    navigator.sendBeacon(this.baseUrl, JSON.stringify(params))
  }
}
