// @guanjia-openclaw/report/dist/index.js
/**
 * @guanjia-openclaw/report
 *
 * 统一埋点上报包
 *
 * 导出统一的 `reportFunc` 方法，自动根据运行环境（浏览器 / Node.js）
 * 选择对应的上报实现（ClientAction / ServerAction）。
 *
 * @example
 * ```ts
 * // 1. 统一方法 - 自动判断环境
 * import { createReporter, reportFunc, REPORT_CONST } from '@guanjia-openclaw/report';
 *
 * // 初始化（任意环境都通用）
 * const reporter = createReporter({ guid: 'xxx', env: 'production' });
 *
 * // 上报（自动选择 ClientAction 或 ServerAction）
 * reporter.reportDirect('click_new', { page_id: 'chat' });
 *
 * // 或使用简便函数
 * reportFunc('click_new', { page_id: 'chat' });
 *
 * // 2. 指定使用服务端（Node.js 环境）
 * import { ServerAction, InstallReporter } from '@guanjia-openclaw/report/server';
 *
 * // 3. 指定使用客户端（浏览器环境）
 * import { ClientAction } from '@guanjia-openclaw/report/client';
 * ```
 */
import { ClientAction } from './client/action.js';
import { ServerAction } from './server/action.js';
// ==================== 环境检测 ====================
/**
 * 检测当前是否为浏览器环境
 */
function isBrowserEnv() {
    // 使用 typeof 检测不会抛出 ReferenceError，但需要显式声明以避免 TS 类型错误
    return typeof globalThis !== 'undefined'
        && 'window' in globalThis
        && 'navigator' in globalThis;
}
// ==================== 统一工厂方法 ====================
/**
 * 创建上报实例（自动检测环境）
 *
 * - 浏览器环境 → 返回 ClientAction
 * - Node.js 环境 → 返回 ServerAction
 *
 * @param options - 上报初始化选项
 * @returns 上报实例，实现 IReportAction 接口
 */
export function createReporter(options = {}) {
    if (isBrowserEnv()) {
        return new ClientAction(options);
    }
    return new ServerAction(options);
}
// ==================== 单例便捷方法 ====================
/** 单例上报实例（延迟创建） */
let _instance = null;
/** 单例初始化选项（由 initReport 设置） */
let _initOptions = {};
/**
 * 初始化全局上报实例
 *
 * 建议在应用启动时调用一次，后续 reportFunc 将复用该实例。
 * 如果不调用 initReport，reportFunc 首次调用时将自动创建默认实例。
 *
 * @param options - 上报初始化选项
 */
export function initReport(options) {
    _initOptions = options;
    // 清除旧实例，下次 reportFunc 调用时重新创建
    _instance = null;
}
/**
 * 获取或创建单例上报实例
 */
function getReporterInstance() {
    if (!_instance) {
        _instance = createReporter(_initOptions);
    }
    return _instance;
}
/**
 * 统一上报方法（自动判断环境）
 *
 * 这是最核心的导出方法。不同端（浏览器渲染进程 / Electron 主进程）
 * 调用相同的方法名，内部自动选择对应的上报实现。
 *
 * @param eventCode - 事件代码（建议使用 REPORT_CONST 中的常量）
 * @param params - 上报参数
 * @param config - 额外配置（可选）
 *
 * @example
 * ```ts
 * import { reportFunc, REPORT_CONST } from '@guanjia-openclaw/report';
 *
 * // 在浏览器中 → 自动使用 ClientAction（fetch + cors）
 * // 在 Node.js 中 → 自动使用 ServerAction（fetch without cors）
 * reportFunc(REPORT_CONST.CLICK_NEW, { page_id: 'chat', action: 'send' });
 * ```
 */
export function reportFunc(eventCode, params = {}, config) {
    if (!eventCode) {
        console.warn('[Report] eventCode 不能为空');
        return;
    }
    try {
        // 兼容 Map 类型参数（与客户端现有调用方式保持兼容）
        let normalized;
        if (params instanceof Map) {
            normalized = {};
            for (const [k, v] of params) {
                normalized[k] = v;
            }
        }
        else {
            normalized = { ...params };
        }
        const reporter = getReporterInstance();
        reporter.reportDirect(eventCode, normalized, config);
    }
    catch (error) {
        console.error(`[Report] 上报失败: ${eventCode}`, error);
    }
}
/**
 * 异步上报方法（可等待上报结果）
 *
 * @param eventCode - 事件代码
 * @param params - 上报参数
 * @param config - 额外配置
 * @returns 上报是否成功
 */
export async function reportFuncAsync(eventCode, params = {}, config) {
    if (!eventCode) {
        console.warn('[Report] eventCode 不能为空');
        return false;
    }
    try {
        let normalized;
        if (params instanceof Map) {
            normalized = {};
            for (const [k, v] of params) {
                normalized[k] = v;
            }
        }
        else {
            normalized = { ...params };
        }
        const reporter = getReporterInstance();
        return await reporter.reportAsync(eventCode, normalized, config);
    }
    catch (error) {
        console.error(`[Report] 异步上报失败: ${eventCode}`, error);
        return false;
    }
}
// 常量
export { REPORT_CONST, DEFAULT_APPKEY, ENV_TYPES, INSTALL_CHANNELS, REPORT_URL_PRODUCTION, REPORT_URL_TEST, REPORT_URL_LOCAL_PROXY, INSTALL_EVENT_FIRST_INSTALL, INSTALL_EVENT_NSIS_START, INSTALL_EVENT_NSIS_FINISH, INSTALL_EVENT_NSIS_CANCEL, } from './constants.js';
// 客户端和服务端 Action（允许直接引用）
export { ClientAction } from './client/action.js';
export { ServerAction } from './server/action.js';
// 注意：InstallReporter 依赖 node:fs/node:path/node:os，不能从主入口导出，
// 否则浏览器端 Vite 构建会报错。请通过子路径导入：
// import { InstallReporter } from '@guanjia-openclaw/report/server'
