// @guanjia-openclaw/report/dist/server/action.js
/**
 * ServerAction - 服务端埋点上报
 *
 * Node.js 环境专用（Electron 主进程 / 安装阶段），
 * 不依赖 import.meta.env、navigator、window 等浏览器 API。
 * 实现 IReportAction 接口，与 ClientAction 保持一致的调用方式。
 */
import { ENV_TYPES, REPORT_URL_PRODUCTION, REPORT_URL_TEST, DEFAULT_APPKEY } from '../constants.js';
export class ServerAction {
    appkey;
    guid;
    onReportSuccess;
    onReportFail;
    baseUrl;
    constructor(options) {
        // 优先使用上层传入的 baseUrl，与 ClientAction 保持一致
        if (!options.baseUrl) {
            throw new Error('[ClientAction] baseUrl 为必填项，请从上层传入上报地址');
        }
        this.baseUrl = options.baseUrl;
        this.appkey = options.appkey ?? DEFAULT_APPKEY;
        this.guid = options.guid ?? '';
        this.onReportSuccess = options.onReportSuccess ?? (() => { });
        this.onReportFail = options.onReportFail ?? (() => { });
    }
    /**
     * 构建上报 payload（与客户端 ClientAction 完全一致的 JSON 格式）
     */
    buildPayload(eventCode, eventParams, config) {
        const normalized = {};
        for (const key in eventParams) {
            const val = eventParams[key];
            normalized[key] = val != null ? String(val) : '';
        }
        normalized['tdbank_imp_date'] = String(Date.now());
        const params = {
            app_key: this.appkey,
            version: '1.0.0',
            guid: this.guid,
            event_code: eventCode,
            params: normalized,
        };
        if (config) {
            for (const key in config) {
                params[key] = config[key];
            }
        }
        return JSON.stringify(params);
    }
    /**
     * 立即上报（fire-and-forget）
     */
    reportDirect(eventCode, eventParams, config) {
        const body = this.buildPayload(eventCode, eventParams, config);
        fetch(this.baseUrl, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body,
        })
            .then((response) => response.json())
            .then((data) => {
            const result = data;
            if (result['retCode'] === 0) {
                this.onReportSuccess();
            }
            else {
                this.onReportFail(result['retCode']);
            }
        })
            .catch(() => this.onReportFail());
    }
    /**
     * 异步上报（返回 Promise）
     */
    async reportAsync(eventCode, eventParams, config) {
        const body = this.buildPayload(eventCode, eventParams, config);
        try {
            const response = await fetch(this.baseUrl, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body,
            });
            const data = (await response.json());
            if (data['retCode'] === 0) {
                this.onReportSuccess();
                return true;
            }
            this.onReportFail(data['retCode']);
            return false;
        }
        catch {
            this.onReportFail();
            return false;
        }
    }
}
